compiler/rustc_type_ir/src/solve.rs

```diff
@@ -211,6 +259,47 @@ pub struct Response<I: Interner> {
     pub external_constraints: I::ExternalConstraints,
 }
 
+/// Additional constraints returned on success.
+#[derive(derivative::Derivative)]
+#[derivative(
+    Clone(bound = ""),
+    Hash(bound = ""),
+    PartialEq(bound = ""),
+    Eq(bound = ""),
+    Debug(bound = ""),
+    Default(bound = "")
+)]
+#[derive(TypeVisitable_Generic, TypeFoldable_Generic)]
+#[cfg_attr(feature = "nightly", derive(HashStable_NoContext))]
+pub struct ExternalConstraintsData<I: Interner> {
+    pub region_constraints: Vec<ty::OutlivesPredicate<I, I::GenericArg>>,
+    pub opaque_types: Vec<(I::LocalDefId, I::GenericArgs, I::Ty)>,
```

Comment:
This is the only debatable thing -- I destructured `OpaqueTypeKey` into `I::LocalDefId` and `I::GenericArgs`. I could uplift `OpaqueTypeKey` I guess, but it seemed like an implementation detail.

Thoughts?

Reply:
I dislike it :< 😅 tuples with 3 elements are questionable

I would prefer uplifting `OpaqueTypeKey` though I feel like maybe we should rename it to `LocalOpaqueTy` or even `LocalAliasTy`?

Reply:
Okie